### DISCUSSION TITLE
Platform-specific XAML - is the current documentation up-to-date for the new solution type?

### DISCUSSION THREAD

**Discussion by darenm** [unoplatform/uno#12009] [Q&A] [answered]
Does the [Platform-specific XAML markup in Uno](https://platform.uno/docs/articles/platform-specific-xaml.html) article reflect the prefixes that can/should be used for the new solution style?

For example, should the documentation include `mobile` and `not_mobile`:

```xml
xmlns:not_mobile="http://schemas.microsoft.com/winfx/2006/xaml/presentation"
xmlns:mobile="http://uno.ui/mobile"
mc:Ignorable="d mobile"
```
Does this work for Windows and WebAssembly targets?

Equally do the `wasm` and `not_wasm` prefixes function as before?

**Comment by Xiaoy312** [comment, marked as answer]
The guide is still up to date. The commonly used platform-specfic xmlnses havent really changed.
In case of doubt, you can always validate against [Uno.UI.SourceGenerators.props](https://github.com/unoplatform/uno/blob/master/src/SourceGenerators/Uno.UI.SourceGenerators/Content/Uno.UI.SourceGenerators.props), just look for the `IncludeXamlNamespaces` and `ExcludeXamlNamespaces` node and their conditions.

---
About `mobile` and `not_mobile` that you might found in [some projects](https://github.com/search?q=org%3Aunoplatform+not_mobile&type=code), these are user-defined specific xmlnses whose implementation can be found below: 
<details>
<summary>click to expand</summary>

https://github.com/unoplatform/uno.toolkit.ui/blob/2.5.5/src/PlatformItemGroups.props#L53-L66
```xml
<Choose>
	<When Condition="$(IsAndroid) or $(IsIOSOrCatalyst)">
		<ItemGroup>
			<IncludeXamlNamespaces Include="mobile" />
			<ExcludeXamlNamespaces Include="not_mobile" />
		</ItemGroup>
	</When>
	<Otherwise>
		<ItemGroup>
			<IncludeXamlNamespaces Include="not_mobile" />
			<ExcludeXamlNamespaces Include="mobile" />
		</ItemGroup>
	</Otherwise>
</Choose>
```
</details>

---
They also work for windows, however, they must be included within the `mc:Ignorable`:
```
xmlns:mc="http://schemas.openxmlformats.org/markup-compatibility/2006"
xmlns:ios="http://uno.ui/ios"
xmlns:android="http://uno.ui/android"
mc:Ignorable="ios android"
```
General rule of thumb is:
- if the xmlns is meant to be included on windows (uwp+winui) then its url should be "http://schemas.microsoft.com/winfx/2006/xaml/presentation" and should not be included in `mc:Ignorable`
- otherwise, it should be included in `mc:Ignorable` and its url can be any text

---
`wasm` and `not_wasm` works, so long as they are used directly (or indirectly as shared project) in projects that doesnt compile into an app, as opposite to a library. The same is true for `skia` and `not_skia` as well. The reason being is that both wasm and skia uses the same `netstandard2.0` target, so when generating the library there no way to know ahead of time the final consumer's target. It could even be both.

**Reply by darenm** [reply]
Thanks for the response. Your comment below is key:

> `wasm` and `not_wasm` works, so long as they are used directly in projects that doesnt compile into a library (as opposite to an app). The same is true for `skia` and `not_skia` as well. The reason being is that both wasm and skia both uses the same netstadnard2.0 target, so when generating the library there no way to know ahead of time the final consumer target.

When using the new style project where the common code is in a library project, that means, by default, they won't work as they used to with the shared project.

A workaround could be to use a static method and `x:Load`, etc:

```xml
<Grid x:Name="grid2" x:Load="{x:Bind root:RuntimeEnvironment.IsNotWasm, Mode=OneWay}">
```

```csharp
    public static class RuntimeEnvironment
    {
        public static bool IsWasm
        {
            
            get 
            {
#if WASM
                return true; 
#else
                return false;
#endif
            }
        }

        public static bool IsNotWasm
        {

            get
            {
#if WASM
                return false; 
#else
                return true;
#endif
            }
        }

    }
```

I could see value in updating the article to include some of the nuance you have mentioned.

**Reply by Youssef1313** [reply]
Note: With .NET 8.0, assuming we get `net8.0-browser` TFM, the `wasm` conditional could be made working with class library projects.
